src/librustc_mir/transform/validate.rs

```diff
@@ -244,6 +248,21 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             );
                         }
                     }
+                    // See what the deaggregator does with arrays. Until the deaggregator processes
+                    // them, we ignore them here.
+                    Rvalue::Aggregate(box AggregateKind::Array { .. }, _) => {}
+                    // Generators exist after the deaggregator, but we eliminate them in the state
+                    // machine transform optimization.
+                    Rvalue::Aggregate(box AggregateKind::Generator(..), _)
+                        if self.mir_phase <= MirPhase::OptimizedWithGenerators => {}
+                    // Until the drop elaboration phase finishes, we still have aggregates.
+                    Rvalue::Aggregate(..) if self.mir_phase <= MirPhase::DropElab => {}
+                    Rvalue::Aggregate(..) => {
+                        self.fail(
+                            location,
+                            format!("{:?} should not exist after the deaggregator has run", rvalue),
+                        );
```

Comment:
This logic is rather hard to follow, I think mostly due to the repeated `Rvalue::Aggregate`. Can you refactor this into just one match arm for `Rvalue::Aggregate`?

Reply:
Might be good to explicitly match on `MirPhase` in that single branch so that if we add more phases, we'll remember to add the correct cases here.